Major DB Schema Changes for Aliases and Taxonomy

jeffdc opened the pull request.

jeffdc commented:
@Megachile @ahiatt14 If you want you can pull this branch and play around with the UI changes related to all of these DB changes. I have not made all of the required changes to the administration side yet so it is not possible to save changes (will crash). This will not fully build yet either, but you can run locally with `yarn dev`.

Things that should work:

* Explore pages - new UI with tree showing Family-Genus-Species
* Global Search - including searching on aliases
* ID page - including selecting a Section (only have data in for Quercus sections)
* Gall detail page - will show aliases
* Host detail page - will show aliases and Section if present
* Family detail page - new UI shows tree of Genus species for the family

Be aware that when running locally the database is also local (comes down with the branch) and changes will not reflect in production. Also please do not add new images as that will upload images to AWS S3 but they will not show up in production and just end up orphaned (I do not yet have a good way to handle this).

ahiatt14 commented:
Will do

Megachile commented:
All of that works for me, although I couldn't test the host aliases because I couldn't think of one I'd assigned any synonyms to. 

It looks like global search already includes aliases in the current version, doesn't it? I can search "ryderi" and get Aceria negundi so it must work.

jeffdc commented:
There are host aliases (both common and scientific) for Coccoloba diversifolia. Currently there is no differentiation in the UI of common vs scientific.

Yes common names/synonyms were included in the old global search as well as the new.

jeffdc commented:
@Megachile An update on progress:

- #38 Data Quality Indicators - done - go to Host and Gall pages as well the admin pages for both
- #41, #61 - species renaming, deprecated names, aliases - done - host  and gall admin pages
- #47 - sections - done - Section Admin screen, also Host Admin, and Host page
- #71 - taxonomy - the read portion is done (so can load on various screens), but Create, Update, Delete not yet working
- #45 - undescribed species - in process
- #44 - many galls per species - in process

If you want to pull down the local branch and play around you can. I only done spotty testing so I am sure stuff is broken. Any feedback is welcome.

jeffdc commented:
@Megachile My current plan is to knock out the stuff for taxonomy, test it, then deploy it all. This means that #44 and #45 will not be part of this deployment. The DB changes to support it all will be there but the functionality will not be in place yet. I am in particular worried about the complexity of #44. This one will require significant thinking about and rework of the gall pages.

jeffdc commented:
I just pushed up a fix for an error on the gall page that prevented it from loading.

Megachile commented:
Sounds good, let us know if you'd like any input or if there's any way we can help!

ahiatt14 commented:
I got it building okay on my machine, but there's a runtime exception I'm trying to puzzle out. Navigating from the homepage to "ID a Gall" I get this:

![Screenshot 2021-03-17 165336](https://user-images.githubusercontent.com/20526218/111597306-68594800-87a4-11eb-9519-336db248cd63.png)

It may be something funny with the build on my end (I already resolved some yarn/node quirk so I wouldn't be surprised).

jeffdc commented:
This was my fault. I failed to make some changes post refactoring. I pushed up changes that should fix this. Also make sure you run `yarn generate` before trying to `yarn dev` to make sure your generated prisma client reflects the latest schema.

Things are unstable right now with this branch. I have some time today to hopefully get it stabilized. So probably better to hold off until I can get that done.

jeffdc commented:
Ok you can try now. I got it all stabilized. I have done little testing but in theory (you know how that works) this should all work now.

jeffdc commented:
By all I mean: #38, #41, #61, #47, #71. I also fixed #99

jeffdc commented:
@ahiatt14 You will get this build error: https://github.com/jeffdc/gallformers/runs/2142766823?check_suite_focus=true

I am going to open a PR against the repo that causes this issue, but for now you can make this change:
line 146:
`editor?: { type: string; options?: [{ value: string; label: string }] };`
to
`editor?: { type: string; options?: { value: string; label: string }[] };`

in:
node_modules/@types/react-bootstrap-table-next/index.d.ts

jeffdc commented:
Opened PR to fix that typing error in the table component: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/51847

ahiatt14 commented:
It all looks good for me now, as far as I've poked around. Thanks a bunch. I'll hand it off to Adam since he knows what to test

jeffdc commented:
@megachile as you poke around I am interested in what you think of how stuff works. Any bugs and stuff are welcome as well. I will be doing testing and fixing tomorrow. Hopefully can get this stabilized for release by Monday or even earlier.

ahiatt14 commented:
Stuff I'm noticing: 

Section doesn't let you submit unless there's something in the Description field, but there's no redirect saying "this is a required field." It is saving sections I'm creating but not associating the species with the section to display on the host page. None of the existing sections show up in general Search either. 

Section search in ID works perfectly. The new section I made shows up in that menu as MISSING GENUS, as if it wanted me to map it to the parent genus. But the page to create it doesn't have a field to do that. It does bring up the associated galls though so apparently the host-section mapping is working, it's just not showing up on the host page (whereas the existing ones added before are showing up just fine). 

In the Add Host page, it's autofilling the genus field on existing species eg in Quercus with Acalypha instead of Quercus. It's also not clearing info from the previously selected species for section and family and alias. The alias will go through to the DB if I hit submit but the genus doesn't seem to. Otherwise the rename tool and alias mapping seem good. Aliases I'm adding are showing up on the host page. 

Data quality indicators on hosts and galls are working for me. Might want to be a little more specific about what those actually indicate though. IE does marking a host "complete" mean that all sources and traits are added for all galls on that species?

#99 is good

Not sure what you did wrt #41 ?

The aliases don't do anything at the moment other than show up in Search and on the gall/host pages right?

In the Add Gall page the family is not autofilling based on genus (and in fact the genus field is not updating at all for me right now, idk if that's a bug or what) but that wasn't something you were planning to implement now, right? Just a future possibility that you've laid the groundwork for?

Megachile commented:
Those are my notes btw if that wasn't obvious, I was just putting them in Alex's account since the build is on his PC.

jeffdc commented:
This is great feedback. I will work through as much of this as I can today as well as do some testing of my own. Hopefully I will have some changes to push up later today.

jeffdc commented:
@ahiatt14 my PR was merged so you can bump the version to 4.0.12 in the package.json and run `yarn install` and it should work now.

https://github.com/DefinitelyTyped/DefinitelyTyped/pull/51847

jeffdc commented:
@megachile 

Section is not in Global Search as we currently do not have a user viewable page that shows a Section. Global search does not search on any taxonomy other than species name (which includes the genus name since it is part of the species name). If we want to add taxonomy to the global search we will need to think more about what that means and how to display results. If this is something you think we need create a new issue and write it up there.

What are you thinking as far as wording for the Data Quality Indicators? There are 4 places:

1) Host Admin
2) Gall Admin
3) Host View
4) Gall View

I believe I have fixed all the other issues. I am going to do more testing then push changes up.

RE #41 the changes are just the Alias stuff. I think that we agreed to not worry too much about all of the source mapping details?

Megachile commented:
Yeah sorry #41 is fine to ignore, I just had it in my head that the aliases were associated with #61 so I wasn't sure if you had done something else related to #41 on top of that. 

As for the quality indicators, that's something we should discuss as a group maybe. I'll make a comment on #38 and post in the Slack.

jeffdc commented:
I am still working on this. I went off on a bit if a tangent re-working the species-source mapping screen to hopefully work better for everyone. It now looks like this:

![Screen Shot 2021-03-22 at 9 56 12 AM](https://user-images.githubusercontent.com/39919/112000973-ded3ae00-8af4-11eb-88da-ab41e4d9177f.png)

The Mapped Source field will pre-populate with any sources that are mapped to the species. To add one simply press the Add New Mapped Source Button and you will get this:

![Screen Shot 2021-03-22 at 9 58 36 AM](https://user-images.githubusercontent.com/39919/112001270-3114cf00-8af5-11eb-8794-25ed71d51a7b.png)

I will add a delete button to delete the selected mapping. 

I thought about using a tables (like aliases) but I think that will not work well since we want a nice big description field to type in.

Thoughts? @Megachile

Megachile commented:
Looks great to me. I've already gotten a ton of mileage out of the link you added for #87 and this will just make it easier.

jeffdc commented:
@Megachile I pushed up all of the changes and you should be able to test out the new species-source mapping screen. I am running through a bunch of testing this AM but AFAIK the bulk of this work is now done.

jeffdc commented:
I am finding bugs and fixing them. Will push up changes in a bit.

jeffdc commented:
### Tested and fixed:

* All of the user-facing (non-admin) pages and they all seem to be working properly
* Admin pages:
    * Glossary - fixed a bug where add/update/delete failed to update the page (data was saved); fixed bug where it was not apparent that URLs was is required 
    * Family - all good
    * Section - all good
    * Host/Gall - fixed bug where Section could not be updated + bug deleting abundance + bug deleting with no images (did not show up to user only backend)
    * Images - no changes so did not test

### Not Tested Yet or Partially Tested:

* Admin:
    * Source
    * Species-Source Mappings -
    * Gall-Host Mappings -

### Known Issues:
* renaming gall fails
* renaming source does not update the name in the list box until refresh

I will keep plugging away at this later tonight and tomorrow but I can not work on it anymore until then. I pushed up the latest.

jeffdc commented:
Finished up my testing and bug fixing. AFAIK it all works. I would really appreciate any and all testing before we push this out to the production environment. There were a lot of changes! I have pushed up all of the changes.

The following items are included in this PR: #38, #41, #61, #47, #71, #72, #92, #98, #99, #102 (back-end work for #44 and #45 was done, but there are no UI changes for those yet).

jeffdc commented on libs/db/gall.ts:
```diff
@@ -152,68 +157,127 @@ export const allGallIds = (): TaskEither<Error, string[]> => {
+    // db.species.findMany({
+    //     select: {
+    //         genus: true,
+    //     },
```
Delete

jeffdc commented on libs/db/speciessource.ts:
```diff
@@ -73,59 +86,60 @@ export const deleteSpeciesSourceByIds = (speciesId: string, sourceId: string): T
+    // // further work, only that it may have failed. If the map in the 2nd line of the pipe is not run the types will be wrong
+    // // in the upsert call. This is surely just my ignorance and there has to be a better way to accomplish this.
+    // // eslint-disable-next-line prettier/prettier
+    // return pipe(
```
delete

jeffdc commented on pages/admin/glossary.tsx:
```diff
@@ -30,150 +28,147 @@ type Props = {
-            setGlossary(glossary.filter((g) => g.id !== id));
-            setDeleteResults(result);
-        };
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
```
delete

jeffdc commented on pages/admin/tester.tsx:
```diff
@@ -0,0 +1,54 @@
+    };
+
+    return (
+        <>
```
Put this behind an authentication.

ahiatt14 commented:
(Adam again:)

DQIs work fine. Only issue is that the new text I wrote for the DQIs was added on the admin side but not on the user side for each gall/host, and IMO that's where it really matters. 

On the Add/edit host page, if I select one host, then click the X to clear or just erase and type a new name, it throws an "Unhandled Runtime Error" TypeError: e[0] is undefined. It doesn't crash the whole thing, refreshing the page fixes it. 

Section page issues all fixed now for me, including the MISSING GENUS thing. 

#92, #98, #99, and #102 all work perfectly. #98 will be great to have especially. My only minor quibble there is that on the Add Mapped Source dropdown, the year/author info is not visible on sources with medium to long titles. Could/should we flip those so that the year/author appears first?

jeffdc commented:
Oops yeah I missed the DQ text on the user's side. Will fix that.

Will also fix that error on the Host page.

Will also figure out a way to fix the issue with long titles on the species-source mapping page.

jeffdc commented:
I actually do not see the host admin error when clearing a previous selection. I pushed up a  [fix](https://github.com/jeffdc/gallformers/pull/86/commits/f703b9615aee1e3c2ef42bbef03e4be2409ecf34) for that after I gave the all clear yesterday so my guess is that the fix did not make into the build you were testing.

HOw is this for the source listing? This is also how it will show up on search results:
![Screen Shot 2021-03-25 at 9 46 43 AM](https://user-images.githubusercontent.com/39919/112483084-150c6a00-8d4f-11eb-8e78-837f605d0d35.png)

Megachile commented:
Looks great to me, thanks!

jeffdc commented:
Ok if we are all good with this I will merge it and then do a deploy tomorrow. It will require some site downtime while I migrate the database. I will do this tomorrow sometime around 10am EDT. I will put a notice on Slack that the site is going down when I take it offline.
